Add frontend support for Manage Users pagination

mpbrown opened the pull request:
# FRONTEND PULL REQUEST

## Related Issue

- Resolves #8103 

## Changes Proposed

- Adds pagination to the Manage Users side nav
- Updates the routing for the default Settings page to use `settings/users/:pageNumber` so that the page number can be retrieved from the url params
- Adds a "No results found" message with a button to clear the search filter

## Testing

- Deployed on dev5
- Select `Access organization account` from the support admin dashboard
- Select `Big Organization`, type in "test" for justification, and click `Access data`
- Select either facility
- Go to the Settings page and it should bring you to Manage Users
- Test that pagination works as expected

## Screenshots

Page 1 of results
![image](https://github.com/user-attachments/assets/53e837ab-e60c-4d69-9fa6-e5b02ef0b470)

Page 5 of results
![image](https://github.com/user-attachments/assets/c672270b-e953-4efc-996b-8eedac4ecfa0)

Last page of results
![image](https://github.com/user-attachments/assets/b5ec8ce8-44a7-4ee8-8ea3-8e5859a8cf22)

Search results with multiple pages
![image](https://github.com/user-attachments/assets/4c3606a6-55cb-4e4c-ba38-dbbf4813c54c)

Search results with single result page
![image](https://github.com/user-attachments/assets/dd0da0b1-d76a-46ba-8e5a-9e2dd3492143)

No results found
![image](https://github.com/user-attachments/assets/012ccb86-f2e8-4a2f-b777-042cf2089e05)

mpbrown commented on backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java:
```diff
@@ -642,8 +642,17 @@ public ManageUsersPageWrapper getPagedUsersAndStatusInCurrentOrg(
+    boolean requestedPageOutOfBounds = startIndex > totalSearchResults;
+
+    if (onlyOnePageOfResults || requestedPageOutOfBounds) {
+      startIndex = 0;
```
Had to handle this edge case because I encountered an error when I was on page 7 of all results and entered a search query. The app requested page 7 of the search results and the backend calculated the start index as a higher index than the end index. This caused an illegal argument error when attempting to create the sublist

mpbrown commented on frontend/src/app/Settings/SettingsNav.tsx:
```diff
@@ -21,7 +21,11 @@ const SettingsNav = () => {
         <li className="usa-nav__secondary-item">
-          <LinkWithQuery to={`/settings`} end className={classNameByActive}>
+          <LinkWithQuery
+            to={`/settings/users/1`}
```
Is there anywhere else we need to update the routing to this settings page? Like possibly from any email templates?

emyl3 replied:
💡 ooo that's a good callout, Mike! Based on a quick search of the static site as well as our email templates in this repo I don't see anything linking to that.
Maybe we should check-in with Jayna or check our email templates in sendgrid to be sure we covered everything?

mpbrown replied:
I checked with Jayna and we confirmed there isn't anything else that needs to be updated for the settings route

mpbrown commented on frontend/src/app/Settings/Users/ManageUsers.tsx:
```diff
@@ -398,109 +413,132 @@ const ManageUsers: React.FC<Props> = ({
-              aria-labelledby={"user-tab-" + user?.id}
-              className="tablet:grid-col padding-left-3 user-detail-column"
-            >
-              <UserHeading
```
This `UserHeading` section got moved below where it is now conditionally shown if the `activeUser` variable has a value. This is because we now have to show an empty state if there are no search results found.

mpbrown commented on frontend/src/app/Settings/Users/UsersSideNav.tsx:
```diff
@@ -7,71 +7,69 @@ import { displayFullName } from "../../utils";
         showSubmitButton={false}
       />
       <nav
-        className="prime-secondary-nav maxh-tablet-lg overflow-y-scroll"
```
Removing scroll bar from the side nav list

emyl3 commented:
One thing I noticed while testing was searching for a user on a previous page, when you have navigated to a later page. The pagination doesn't reset. Hope this video makes more sense of what I'm trying to say 😅 

https://github.com/user-attachments/assets/19d4c333-1aa2-443d-98fc-3c2a31d42c16

mpbrown commented:
> One thing I noticed while testing was searching for a user on a previous page, when you have navigated to a later page. The pagination doesn't reset. Hope this video makes more sense of what I'm trying to say 😅

Hmm I'll check with product and design to see what the intended behavior should be in this scenario. Right now the component determines the current page based on the url parameter so in that example the url param would still be 7 even when the search results returned just one page. We could possibly have the frontend trigger a redirect to the 1st page whenever a search is entered. I remember we discovered that focus redirect bug on the results page when we were pairing on this

khayeni-SL commented:
In a (probably unlikely) scenario where the results return too many patient names to fit on one page, how would the user navigate to "page 2"? (Ex: There are 20+ Wallaces.) Is there a large lift to have the pagination reflect the results filter?

Alternatively, I'd want to know if pagination could be disabled so it does not imply the user can navigate between pages if the filter is active. cc @kenieh for thoughts

mpbrown commented:
> In a (probably unlikely) scenario where the results return too many patient names to fit on one page, how would the user navigate to "page 2"? (Ex: There are 20+ Wallaces.) Is there a large lift to have the pagination reflect the results filter?

Yes this type of paginated search should already be supported! So if you were to search by a single character like "a" it'll show multiple pages of filtered results

mpbrown commented:
@khayeni-SL @kenieh I updated dev5 with the styling changes and fixed the bug @emyl3 found. The search is now handled with the url query parameter `name` so when a search is input it will always bring them to the first page of those search results. PR should be all ready for review!

emyl3 commented on frontend/src/app/Settings/Settings.tsx:
```diff
@@ -29,7 +29,7 @@ const Settings = () => {
             element={<ManageSelfRegistrationLinksContainer />}
           />
-          <Route path="/" element={<ManageUsersContainer />} />
+          <Route path="users/:pageNumber" element={<ManageUsersContainer />} />
```
I forgot we could also add something [like the following](https://reactrouter.com/6.29.0/route/route#index):
```
<Route index element={<Navigate replace to="users/1"/>}/>
```
to default it to the users page like before if someone were to navigate to `/settings`

mpbrown replied:
Good idea! I just added this redirect

mpbrown replied:
Updated the urls we have configured for the Lighthouse checks since adding this new redirect confused it

mpbrown replied:
I wonder if that's because we'd have to add the active facility id to the search params like we do in the ManageUsersContainer component. I can test that out and see if it works

```
let searchParams: Record<string, string> = {
      facility: activeFacilityId,
    };
...
navigate({
      pathname: "/settings/users/1",
      search: new URLSearchParams(searchParams).toString(),
    });
```

mpbrown replied:
@emyl3 I'm redeploying to dev5 with a fix for this, although it will still prompt the user to select which facility but only once

emyl3 replied:
ooo wow ok! Thank you for that quick fix. That prompt happening once totally makes sense!

emyl3 reviewed (approved):
Left a couple nits and one thing I noticed based on our [search implementation](https://github.com/CDCgov/prime-simplereport/pull/8371/files#diff-3a4cd84e094d22aa2350c1d2c5c0a1a66e1a4952f73b13ef86654c7e2d285477R631) (Sorry I didn't notice this earlier when you introduced the implementation. 😓)

Looks like this search behaves slightly differently than our other ones.

https://github.com/user-attachments/assets/3db74dc6-7ad7-47f5-8fa8-e60bc58f4321

Let me know what you decide to do. Everything else looks good and works as expected for me.

mpbrown commented:
> Looks like this search behaves slightly differently than our other ones.

This is intended since we want the user search results to appear in that side nav panel

mpbrown commented:
> > > Looks like this search behaves slightly differently than our other ones.
> > 
> > 
> > This is intended since we want the user search results to appear in that side nav panel @emyl3
> 
> Just to clarify, I was talking about searching for a user based on "Last Name, First Name" that works in our other searches but not this one.

Oh whoops I thought the clip was about the dropdown section for selecting the filter! Good callout I'll check with product to confirm for supporting the other search format as well

bobbywells52 commented on frontend/src/app/Settings/Users/ManageUsers.test.tsx:
```diff
@@ -362,71 +363,20 @@ describe("ManageUsers", () => {
             updateUserName={updateUserName}
             updateUserEmail={updateUserEmail}
+            currentPage={1}
+            entriesPerPage={10}
```
Nit: should we have this value set to 12 as that's the [constant that is set for Okta Page Size](https://github.com/CDCgov/prime-simplereport/pull/8371/files#diff-bce10eb70959e07794e82fcc4ef0f8cd1c7d002430ee95d5bf0bf66c75c82656R19)?

bobbywells52 reviewed (approved):
Tested in dev5 and everything is working as expected! I've left a few non-blocking nits. Thanks for your great work on this Mike!

mpbrown commented:
> Non blocking nit: The styling and spacing for the page toggles is inconsistent between the 3 states of first page, any middle page, and last page. Screenshots below:

I believe this is intentional so there is not left padding of white space on the first page, but tagging @kenieh to confirm

mpbrown commented:
@emyl3 I updated the search filtering so now it should handle more flexible search queries like the results page does, thanks for finding that!

arinkulshi-skylight reviewed (approved):
I have tested the changes in dev5. LGTM!

emyl3 reviewed (dismissed):
LGTM! Left a comment about the redirect fix I initially proposed 😓 Everything else works as expected!! Thank you for all your work on this.

emyl3 commented on frontend/src/app/Settings/Settings.tsx:
```diff
@@ -10,6 +12,10 @@ import { ManageSelfRegistrationLinksContainer } from "./ManageSelfRegistrationLi
 const Settings = () => {
+  const [facility] = useSelectedFacility();
+  const activeFacilityId = facility?.id || "";
+  const settingsIndexRedirect = "/settings/users/1?" + activeFacilityId;
```
Don't we need the "/settings/users/1?facility=" + activeFacilityId here? (specifically adding `facility=`?

mpbrown replied:
😬 *face palm* yes we definitely should need that here! Somehow the redirect still worked but that's probably why it's still hitting that other facility select screen. Thank you for catching this!

emyl3 reviewed (approved):
checked that change locally! and retested the other functionality on dev5 -- LGTM!!! Thank you, Mike!!! 🤩

mpbrown commented:
Design final approval from @kenieh on [Slack linked here](https://skylight-hq.slack.com/archives/C051X0Q3HGF/p1739299829496219?thread_ts=1739290590.945859&cid=C051X0Q3HGF) ✅

bobbywells52 reviewed (approved):
🚢 🚀  thanks for the big lift on this Mike!
